In [1]:
import numpy as np
import pandas as pd
import random
from faker import Faker

In [2]:
#initialize Faker
fake=Faker()

In [3]:
names = []
n_names=500000
for n in range(n_names):
    names.append(fake.name())

In [4]:
customer_segment = ['High Value','Medium Value', 'Low Value']

In [5]:
customer_segments=np.random.choice(customer_segment,n_names, p=[.20,.25,.55])

In [6]:
user_id =[]
for i in range(n_names):
    user_id.append(np.random.randint(1,5000))

In [7]:
variables=[user_id,names,customer_segments]

df=pd.DataFrame(variables).transpose()

df.columns=["User ID", "Customer Name","customer_segment"]

In [8]:
df.head()

User ID       Customer Name customer_segment
0     283    Christine Stuart        Low Value
1    2319      Brittany Price       High Value
2    3761        Travis James       High Value
3    1280  Alexandra Mcdonald        Low Value
4     984     Katherine Simon        Low Value

In [9]:
df.describe()

User ID  Customer Name customer_segment
count    500000         500000           500000
unique     4999         221788                3
top        4294  Michael Smith        Low Value
freq        140            224           274977

In [10]:
### regions
regions=["HN","HS","MAT","MVO","Mash E+CZA","MID","MAN","Mash C+W"]
df["Region"]=np.random.choice(regions,n_names)

In [11]:
### products
products_list = ['Private WiFi Bundle 75GB', 'Data Daily Bundle 2GB', 'Private WiFi Bundle 50GB',
       'Daily Data Bouquet 2GB', 'Private WiFi Bundle  15GB', 'Private WiFi Bundle 25GB', 
       '2hr data bundle 1.5GB', 'eLearning Bundle 10GB $1120', 'WhatsApp (400MB)+ Pinterest 200MB + 400MB Sasai Bouquet',
       'WhatsApp + Twitter Monthly Bundle $3', 'Private WiFi Bundle 8GB', 'Data Weekly Bundle 700MB', 'eLearning Bundle 10GB $1120',
       'WhatsApp (400MB)+ Pinterest 200MB + 400MB Sasai Bouquet', 'Data Weekly Bundle 350MB', '2hr data bundle 1.5GB', 
       'Instagram (400MB) + 400MB Sasai Bouquet', 'Daily Data Bouquet 250MB', 'Daily Data Bouquet 40MB','Data Daily Bundle 40MB',
       'Twitter (45MB)+ LinkedIn 25MB + 45MB Sasai Bouquet','Facebook + Instagram Weekly Bundle $1',
       'Twitter (140MB)+ LinkedIn 70MB + 140MB Sasai Bouquet','Twitter Monthly Balance 400MB', 'Data Weekly Bouquet 160MB', 'WhatsApp + Twitter Weekly Bundle $1', 'WhatsApp (400MB)+ Pinterest 200MB + 400MB Sasai Bouquet',
       'Twitter (140MB)+ LinkedIn 70MB + 140MB Sasai Bouquet', 'Data Weekly Bundle 350MB', 'Daily Data Bouquet 80MB',
       'WhatsApp (45MB)+ Pinterest 25MB + 45MB Sasai Bouquet', 'WhatsApp (20MB)+ Pinterest 10MB + 20MB Sasai Bouquet',
       'Data Daily Bundle 600MB', 'Data Weekly Bundle 300MB', 'Facebook (45MB)+ SnapChat 25MB + 45MB Sasai Bouquet',
       '1hr data bundle 1GB', 'Daily Data Bouquet 20MB', 'Instagram (20MB) + 20MB Sasai Bouquet', 'Facebook + Instagram Daily Bundle $0.3'
       ]

In [12]:
### removing duplicates
products_list = list(dict.fromkeys(products_list))

In [13]:
len(products_list)

33

In [14]:
df["data_subscriptions"]=np.random.choice(products_list,n_names)

In [15]:
### data volume
df["Data_volume"]=0

for i in range(len(df)):
    if df["customer_segment"][i]=="High Value":
        df["Data_volume"][i]=np.random.randint(29860,115000)
    elif df["customer_segment"][i]=="Medium Value":
        df["Data_volume"][i]=np.random.randint(13694,24531)
    else:
        df["Data_volume"][i]=np.random.randint(1,6741)

C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\ipykernel_launcher.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  # Remove the CWD from sys.path while we load stuff.
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\ipykernel_launcher.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://

In [22]:
df.head()

User ID       Customer Name customer_segment Region  \
0     283    Christine Stuart        Low Value    MID   
1    2319      Brittany Price       High Value    MAN   
2    3761        Travis James       High Value    MID   
3    1280  Alexandra Mcdonald        Low Value    MAT   
4     984     Katherine Simon        Low Value    MVO   

                                  data_subscriptions  Data_volume  
0                                1hr data bundle 1GB         1202  
1  Facebook (45MB)+ SnapChat 25MB + 45MB Sasai Bo...        89348  
2                            Daily Data Bouquet 20MB       110723  
3                           Data Weekly Bundle 350MB         2456  
4  Twitter (45MB)+ LinkedIn 25MB + 45MB Sasai Bou...          559

In [24]:
df.describe(include=object)

User ID  Customer Name customer_segment  Region     data_subscriptions
count    500000         500000           500000  500000                 500000
unique     4999         221788                3       8                     33
top        4294  Michael Smith        Low Value      HN  Data Daily Bundle 2GB
freq        140            224           274977   62736                  15391

In [25]:
### bringing in the products and their keys
productsdf = pd.read_csv('C:\\Users\\trevor.muchenje\\Desktop\\disser\\products_keys.csv')
productsdf.head()

product_id                            product_name
0           6  Facebook + Instagram Monthly Bundle $3
1           8                Private WiFi Bundle 75GB
2          10             WhatsApp Daily Balance 20MB
3          11                   Data Daily Bundle 2GB
4          12          Instagram Weekly Balance 140MB

In [26]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500000 entries, 0 to 499999
Data columns (total 6 columns):
 #   Column              Non-Null Count   Dtype 
---  ------              --------------   ----- 
 0   User ID             500000 non-null  object
 1   Customer Name       500000 non-null  object
 2   customer_segment    500000 non-null  object
 3   Region              500000 non-null  object
 4   data_subscriptions  500000 non-null  object
 5   Data_volume         500000 non-null  int64 
dtypes: int64(1), object(5)
memory usage: 22.9+ MB


In [27]:
cdf=df.merge(productsdf, left_on='data_subscriptions',right_on='product_name' , how='left')

In [28]:
cdf.head()

User ID       Customer Name customer_segment Region  \
0     283    Christine Stuart        Low Value    MID   
1    2319      Brittany Price       High Value    MAN   
2    3761        Travis James       High Value    MID   
3    1280  Alexandra Mcdonald        Low Value    MAT   
4     984     Katherine Simon        Low Value    MVO   

                                  data_subscriptions  Data_volume  product_id  \
0                                1hr data bundle 1GB         1202         263   
1  Facebook (45MB)+ SnapChat 25MB + 45MB Sasai Bo...        89348         255   
2                            Daily Data Bouquet 20MB       110723         262   
3                           Data Weekly Bundle 350MB         2456         197   
4  Twitter (45MB)+ LinkedIn 25MB + 45MB Sasai Bou...          559         182   

                                        product_name  
0                                1hr data bundle 1GB  
1  Facebook (45MB)+ SnapChat 25MB + 45MB Sasai Bo...  
2                            Daily Data Bouquet 20MB  
3                           Data Weekly Bundle 350MB  
4  Twitter (45MB)+ LinkedIn 25MB + 45MB Sasai Bou...

In [29]:
cdf.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 500000 entries, 0 to 499999
Data columns (total 8 columns):
 #   Column              Non-Null Count   Dtype 
---  ------              --------------   ----- 
 0   User ID             500000 non-null  object
 1   Customer Name       500000 non-null  object
 2   customer_segment    500000 non-null  object
 3   Region              500000 non-null  object
 4   data_subscriptions  500000 non-null  object
 5   Data_volume         500000 non-null  int64 
 6   product_id          500000 non-null  int64 
 7   product_name        500000 non-null  object
dtypes: int64(2), object(6)
memory usage: 34.3+ MB


In [30]:
dff = cdf[['User ID','customer_segment','Region', 'product_id','data_subscriptions', 'Data_volume']]

In [31]:
dff.head()

User ID customer_segment Region  product_id  \
0     283        Low Value    MID         263   
1    2319       High Value    MAN         255   
2    3761       High Value    MID         262   
3    1280        Low Value    MAT         197   
4     984        Low Value    MVO         182   

                                  data_subscriptions  Data_volume  
0                                1hr data bundle 1GB         1202  
1  Facebook (45MB)+ SnapChat 25MB + 45MB Sasai Bo...        89348  
2                            Daily Data Bouquet 20MB       110723  
3                           Data Weekly Bundle 350MB         2456  
4  Twitter (45MB)+ LinkedIn 25MB + 45MB Sasai Bou...          559

In [25]:
dff.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 500000 entries, 0 to 499999
Data columns (total 6 columns):
 #   Column              Non-Null Count   Dtype 
---  ------              --------------   ----- 
 0   User ID             500000 non-null  object
 1   customer_segment    500000 non-null  object
 2   Region              500000 non-null  object
 3   product_id          500000 non-null  int64 
 4   data_subscriptions  500000 non-null  object
 5   Data_volume         500000 non-null  int64 
dtypes: int64(2), object(4)
memory usage: 26.7+ MB


In [33]:
dff.to_excel('data.xlsx')

### Modelling

In [26]:
import sys
import scipy.sparse as sparse
from scipy.sparse.linalg import spsolve
import random
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
import implicit
from pandas.api.types import CategoricalDtype

In [27]:
# Before we make any sort of ratings matrix, it would be nice to have a lookup table that keeps track of each
# item ID along with a description of that item. Let's make that now.

In [28]:
item_lookup = dff[['product_id', 'data_subscriptions']].drop_duplicates() # Only get unique item/description pairs
#item_lookup['ProductKey'] = item_lookup.ProductKey.astype(str) # Encode as strings for future lookup ease
item_lookup['product_id'] = item_lookup.product_id.astype(int)

In [29]:
item_lookup.head()

product_id                                 data_subscriptions
0         125                             Data Daily Bundle 40MB
1         258  WhatsApp (400MB)+ Pinterest 200MB + 400MB Sasa...
2         140                            Daily Data Bouquet 40MB
3         256              Instagram (20MB) + 20MB Sasai Bouquet
5         179  WhatsApp (20MB)+ Pinterest 10MB + 20MB Sasai B...

In [30]:
item_lookup.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 33 entries, 0 to 105
Data columns (total 2 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   product_id          33 non-null     int32 
 1   data_subscriptions  33 non-null     object
dtypes: int32(1), object(1)
memory usage: 660.0+ bytes


In [31]:
dff.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 500000 entries, 0 to 499999
Data columns (total 6 columns):
 #   Column              Non-Null Count   Dtype 
---  ------              --------------   ----- 
 0   User ID             500000 non-null  object
 1   customer_segment    500000 non-null  object
 2   Region              500000 non-null  object
 3   product_id          500000 non-null  int64 
 4   data_subscriptions  500000 non-null  object
 5   Data_volume         500000 non-null  int64 
dtypes: int64(2), object(4)
memory usage: 26.7+ MB


In [32]:
# Group purchase quantities together by ProductKey and item ID Change any sums that equal zero to one 
# (this can happen if items were returned, but we want to indicate that the user actually purchased the item instead of 
#  assuming no interaction between the user and the item ever took place) Only include customers with a positive purchase 
# total to eliminate possible errors Set up our sparse ratings matrix This last step is especially important if you don't 
# want to have unnecessary memory issues! If you think about it, our matrix is going to contain thousands of items and 
# thousands of users with a user/item value required for every possible combination. That is a LARGE matrix, so we can 
# save a lot of memory by keeping the matrix sparse and only saving the locations and values of items that are not zero.

# The code below will finish the preprocessing steps necessary for our final ratings sparse matrix:

In [33]:
dff['User ID'] = dff['User ID'].astype(int) # Convert to int for userid
dff = dff[['product_id', 'Data_volume', 'User ID']] # Get rid of unnecessary info
grouped_cleaned = dff.groupby(['User ID', 'product_id']).sum().reset_index() # Group together
grouped_cleaned.Data_volume.loc[grouped_cleaned.Data_volume == 0] = 1 # Replace a sum of zero purchases with a one to
# indicate purchased
grouped_purchased = grouped_cleaned.query('Data_volume > 0') # Only get customers where purchase totals were positive

C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [34]:
grouped_purchased.head()

User ID  product_id  Data_volume
0        1           8        21639
1        1          11       180746
2        1         118        11234
3        1         122        23909
4        1         125       133066

In [35]:
grouped_purchased.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 156983 entries, 0 to 156982
Data columns (total 3 columns):
 #   Column       Non-Null Count   Dtype
---  ------       --------------   -----
 0   User ID      156983 non-null  int64
 1   product_id   156983 non-null  int64
 2   Data_volume  156983 non-null  int64
dtypes: int64(3)
memory usage: 4.8 MB


In [36]:
# Instead of representing an explicit rating, the data volume can represent a "confidence" in terms of how strong the
# interaction was. Items with a larger number of data volumes by a customer can carry more weight in our ratings matrix of
# purchases.

# Our last step is to create the sparse ratings matrix of users and items utilizing the code below:

In [37]:
customers = list(np.sort(grouped_purchased['User ID'].unique())) # Get our unique customers
products = list(grouped_purchased.product_id.unique()) # Get our unique products that were purchased
quantity = list(grouped_purchased.Data_volume) # All of our purchases

cat_type1 = CategoricalDtype(categories=customers)
rows = grouped_purchased['User ID'].astype(cat_type1).cat.codes 
# Get the associated row indices

cat_type2 = CategoricalDtype(categories=products)
cols = grouped_purchased.product_id.astype(cat_type2).cat.codes 
# Get the associated column indices

purchases_sparse = sparse.csr_matrix((quantity, (rows, cols)), shape=(len(customers), len(products)))

In [38]:
# df = df['Grades'].astype('category',categories=['D', 'D+', 'C-', 'C', 'C+', 'B-', 'B', 'B+', 'A-', 'A', 'A+'],ordered=True)

# cat_type = CategoricalDtype(categories=cats, ordered=True)
# df['Grades'] = df['Grades'].astype(cat_type)

In [39]:
purchases_sparse

<4999x33 sparse matrix of type '<class 'numpy.intc'>'
	with 156983 stored elements in Compressed Sparse Row format>

##### We have 4999 customers with 33 items. For these user/item interactions, 157075 of these items had a purchase.

In [40]:
matrix_size = purchases_sparse.shape[0]*purchases_sparse.shape[1] # Number of possible interactions in the matrix
num_purchases = len(purchases_sparse.nonzero()[0]) # Number of items interacted with
sparsity = 100*(1 - (num_purchases/matrix_size))
sparsity

4.839755829953873

###### 4.78% of the interaction matrix is sparse. For collaborative filtering to work, the maximum sparsity you could get away with would probably be about 99.5% or so. 
###### We are well below this, so we should be able to get decent results.

## Creating a Training and Validation Set

Typically in Machine Learning applications, we need to test whether the model we just trained is any good on new data it hasn't yet seen before from the training phase. We do this by creating a test set completely separate from the training set. Usually this is fairly simple: just take a random sample of the training example rows in our feature matrix and separate it away from the training set. That normally looks like this:

With collaborative filtering, that's not going to work because you need all of the user/item interactions to find the proper matrix factorization. A better method is to hide a certain percentage of the user/item interactions from the model during the training phase chosen at random. Then, check during the test phase how many of the items that were recommended the user actually ended up purchasing in the end. Ideally, you would ultimately test your recommendations with some kind of A/B test or utilizing data from a time series where all data prior to a certain point in time is used for training while data after a certain period of time is used for testing.

For this example, because the time period is only 8 months and because of the purchasing type (products), it is most likely products won't be purchased again in a short time period anyway. This will be a better test. You can see an example here:

Our test set is an exact copy of our original data. The training set, however, will mask a random percentage of user/item interactions and act as if the user never purchased the item (making it a sparse entry with a zero). We then check in the test set which items were recommended to the user that they ended up actually purchasing. If the users frequently ended up purchasing the items most recommended to them by the system, we can conclude the system seems to be working.

As an additional check, we can compare our system to simply recommending the most popular items to every user (beating popularity is a bit difficult). This will be our baseline.

This method of testing isn't necessarily the "correct" answer, because it depends on how you want to use the recommender system. However, it is a practical way of testing performance I will use for this example.

Now that we have a plan on how to separate our training and testing sets, let's create a function that can do this for us. We will also import the random library and set a seed so that you will see the same results as I did

In [41]:
import random

In [42]:
def make_train(ratings, pct_test = 0.2):
    '''
    This function will take in the original user-item matrix and "mask" a percentage of the original ratings where a
    user-item interaction has taken place for use as a test set. The test set will contain all of the original ratings, 
    while the training set replaces the specified percentage of them with a zero in the original ratings matrix. 
    
    parameters: 
    
    ratings - the original ratings matrix from which you want to generate a train/test set. Test is just a complete
    copy of the original set. This is in the form of a sparse csr_matrix. 
    
    pct_test - The percentage of user-item interactions where an interaction took place that you want to mask in the 
    training set for later comparison to the test set, which contains all of the original ratings. 
    
    returns:
    
    training_set - The altered version of the original data with a certain percentage of the user-item pairs 
    that originally had interaction set back to zero.
    
    test_set - A copy of the original ratings matrix, unaltered, so it can be used to see how the rank order 
    compares with the actual interactions.
    
    user_inds - From the randomly selected user-item indices, which user rows were altered in the training data.
    This will be necessary later when evaluating the performance via AUC.
    '''
    test_set = ratings.copy() # Make a copy of the original set to be the test set. 
    test_set[test_set != 0] = 1 # Store the test set as a binary preference matrix
    training_set = ratings.copy() # Make a copy of the original data we can alter as our training set. 
    nonzero_inds = training_set.nonzero() # Find the indices in the ratings data where an interaction exists
    nonzero_pairs = list(zip(nonzero_inds[0], nonzero_inds[1])) # Zip these pairs together of user,item index into list
    random.seed(0) # Set the random seed to zero for reproducibility
    num_samples = int(np.ceil(pct_test*len(nonzero_pairs))) # Round the number of samples needed to the nearest integer
    samples = random.sample(nonzero_pairs, num_samples) # Sample a random number of user-item pairs without replacement
    user_inds = [index[0] for index in samples] # Get the user row indices
    item_inds = [index[1] for index in samples] # Get the item column indices
    training_set[user_inds, item_inds] = 0 # Assign all of the randomly chosen user-item pairs to zero
    training_set.eliminate_zeros() # Get rid of zeros in sparse array storage after update to save space
    return training_set, test_set, list(set(user_inds)) # Output the unique list of user rows that were altered

This will return our training set, a test set that has been binarized to 0/1 for purchased/not purchased, and a list of which users had at least one item masked. We will test the performance of the recommender system on these users only. I am masking 20% of the user/item interactions for this example.

In [43]:
product_train, product_test, product_users_altered = make_train(purchases_sparse, pct_test = 0.2)

Now that we have our train/test split, it is time to implement the alternating least squares algorithm from the Hu, Koren, and Volinsky paper.

## Implementing ALS for Implicit Feedback

In [44]:
def implicit_weighted_ALS(training_set, lambda_val = 0.1, alpha = 40, iterations = 10, rank_size = 20, seed = 0):
    '''
    Implicit weighted ALS taken from Hu, Koren, and Volinsky 2008. Designed for alternating least squares and implicit
    feedback based collaborative filtering. 
    
    parameters:
    
    training_set - Our matrix of ratings with shape m x n, where m is the number of users and n is the number of items.
    Should be a sparse csr matrix to save space. 
    
    lambda_val - Used for regularization during alternating least squares. Increasing this value may increase bias
    but decrease variance. Default is 0.1. 
    
    alpha - The parameter associated with the confidence matrix discussed in the paper, where Cui = 1 + alpha*Rui. 
    The paper found a default of 40 most effective. Decreasing this will decrease the variability in confidence between
    various ratings.
    
    iterations - The number of times to alternate between both user feature vector and item feature vector in
    alternating least squares. More iterations will allow better convergence at the cost of increased computation. 
    The authors found 10 iterations was sufficient, but more may be required to converge. 
    
    rank_size - The number of latent features in the user/item feature vectors. The paper recommends varying this 
    between 20-200. Increasing the number of features may overfit but could reduce bias. 
    
    seed - Set the seed for reproducible results
    
    returns:
    
    The feature vectors for users and items. The dot product of these feature vectors should give you the expected 
    "rating" at each point in your original matrix. 
    '''
    
    # first set up our confidence matrix
    
    conf = (alpha*training_set) # To allow the matrix to stay sparse, I will add one later when each row is taken 
                                # and converted to dense. 
    num_user = conf.shape[0]
    num_item = conf.shape[1] # Get the size of our original ratings matrix, m x n
    
    # initialize our X/Y feature vectors randomly with a set seed
    rstate = np.random.RandomState(seed)
    
    X = sparse.csr_matrix(rstate.normal(size = (num_user, rank_size))) # Random numbers in a m x rank shape
    Y = sparse.csr_matrix(rstate.normal(size = (num_item, rank_size))) # Normally this would be rank x n but we can 
                                                                 # transpose at the end. Makes calculation more simple.
    X_eye = sparse.eye(num_user)
    Y_eye = sparse.eye(num_item)
    lambda_eye = lambda_val * sparse.eye(rank_size) # Our regularization term lambda*I. 
    
    # We can compute this before iteration starts. 
    
    # Begin iterations
   
    for iter_step in range(iterations): # Iterate back and forth between solving X given fixed Y and vice versa
        # Compute yTy and xTx at beginning of each iteration to save computing time
        yTy = Y.T.dot(Y)
        xTx = X.T.dot(X)
        # Being iteration to solve for X based on fixed Y
        for u in range(num_user):
            conf_samp = conf[u,:].toarray() # Grab user row from confidence matrix and convert to dense
            pref = conf_samp.copy() 
            pref[pref != 0] = 1 # Create binarized preference vector 
            CuI = sparse.diags(conf_samp, [0]) # Get Cu - I term, which is just CuI since we never added 1
            yTCuIY = Y.T.dot(CuI).dot(Y) # This is the yT(Cu-I)Y term 
            yTCupu = Y.T.dot(CuI + Y_eye).dot(pref.T) # This is the yTCuPu term, where we add the eye back in
                                                      # Cu - I + I = Cu
            X[u] = spsolve(yTy + yTCuIY + lambda_eye, yTCupu) 
            # Solve for Xu = ((yTy + yT(Cu-I)Y + lambda*I)^-1)yTCuPu, equation 4 from the paper  
        # Begin iteration to solve for Y based on fixed X 
        for i in range(num_item):
            conf_samp = conf[:,i].T.toarray() # transpose to get it in row format and convert to dense
            pref = conf_samp.copy()
            pref[pref != 0] = 1 # Create binarized preference vector
            CiI = sparse.diags(conf_samp, [0]) # Get Ci - I term, which is just CiI since we never added 1
            xTCiIX = X.T.dot(CiI).dot(X) # This is the xT(Cu-I)X term
            xTCiPi = X.T.dot(CiI + X_eye).dot(pref.T) # This is the xTCiPi term
            Y[i] = spsolve(xTx + xTCiIX + lambda_eye, xTCiPi)
            # Solve for Yi = ((xTx + xT(Cu-I)X) + lambda*I)^-1)xTCiPi, equation 5 from the paper
    # End iterations
    return X, Y.T # Transpose at the end to make up for not being transposed at the beginning. 
                         # Y needs to be rank x n. Keep these as separate matrices for scale reasons.

Hopefully the comments are enough to see how the code was structured. You want to keep the matrices sparse where possible to avoid memory issues! Let's try just a single iteration of the code to see how it works (it's pretty slow right now!) I will choose 20 latent factors as my rank matrix size along with an alpha of 15 and regularization of 0.1 (which I found in testing does the best).

In [45]:
user_vecs, item_vecs = implicit_weighted_ALS(product_train, lambda_val = 0.1, alpha = 15, iterations = 10,rank_size = 20)

We can investigate ratings for a particular user by taking the dot product between the user and item vectors (U and V). Let's look at our first user.

In [46]:
user_vecs[0,:].dot(item_vecs).toarray()[0,:5]

array([1.00107033, 1.00046134, 1.00059622, 0.98557576, 0.99998503])

#### Speeding Up ALS

In [47]:
alpha = 15
user_vecs, item_vecs = implicit.alternating_least_squares((product_train*alpha).astype('double'), 
                                                          factors=20, 
                                                          regularization = 0.1, 
                                                         iterations = 50)

This method is deprecated. Please use the AlternatingLeastSquares class instead


  0%|          | 0/50 [00:00<?, ?it/s]

#### Evaluating the Recommender System

Remember that our training set had 20% of the purchases masked? This will allow us to evaluate the performance of our recommender system. Essentially, we need to see if the order of recommendations given for each user matches the items they ended up purchasing. A commonly used metric for this kind of problem is the area under the Receiver Operating Characteristic (or ROC) curve. A greater area under the curve means we are recommending items that end up being purchased near the top of the list of recommended items. Usually this metric is used in more typical binary classification problems to identify how well a model can predict a positive example vs. a negative one. It will also work well for our purposes of ranking recommendations.

In order to do that, we need to write a function that can calculate a mean area under the curve (AUC) for any user that had at least one masked item. As a benchmark, we will also calculate what the mean AUC would have been if we had simply recommended the most popular items. Popularity tends to be hard to beat in most recommender system problems, so it makes a good comparison.

First, let's make a simple function that can calculate our AUC. Scikit-learn has one we can alter a bit.

In [48]:
from sklearn import metrics

In [49]:
def auc_score(predictions, test):
    '''
    This simple function will output the area under the curve using sklearn's metrics. 
    
    parameters:
    
    - predictions: your prediction output
    
    - test: the actual target result you are comparing to
    
    returns:
    
    - AUC (area under the Receiver Operating Characterisic curve)
    '''
    fpr, tpr, thresholds = metrics.roc_curve(test, predictions)
    return metrics.auc(fpr, tpr)

Now, utilize this helper function inside of a second function that will calculate the AUC for each user in our training set that had at least one item masked. It should also calculate AUC for the most popular items for our users to compare.

In [50]:
def calc_mean_auc(training_set, altered_users, predictions, test_set):
    '''
    This function will calculate the mean AUC by user for any user that had their user-item matrix altered. 
    
    parameters:
    
    training_set - The training set resulting from make_train, where a certain percentage of the original
    user/item interactions are reset to zero to hide them from the model 
    
    predictions - The matrix of your predicted ratings for each user/item pair as output from the implicit MF.
    These should be stored in a list, with user vectors as item zero and item vectors as item one. 
    
    altered_users - The indices of the users where at least one user/item pair was altered from make_train function
    
    test_set - The test set constucted earlier from make_train function
    
    
    
    returns:
    
    The mean AUC (area under the Receiver Operator Characteristic curve) of the test set only on user-item interactions
    there were originally zero to test ranking ability in addition to the most popular items as a benchmark.
    '''
    
    
    store_auc = [] # An empty list to store the AUC for each user that had an item removed from the training set
    popularity_auc = [] # To store popular AUC scores
    
    pop_items = np.array(test_set.sum(axis = 0)).reshape(-1) # Get sum of item iteractions to find most popular
    item_vecs = predictions[1]
    for user in altered_users: # Iterate through each user that had an item altered
        training_row = training_set[user,:].toarray().reshape(-1) # Get the training set row
        zero_inds = np.where(training_row == 0) # Find where the interaction had not yet occurred
        # Get the predicted values based on our user/item vectors
        user_vec = predictions[0][user,:]
        pred = user_vec.dot(item_vecs).toarray()[0,zero_inds].reshape(-1)
        # Get only the items that were originally zero
        # Select all ratings from the MF prediction for this user that originally had no iteraction
        actual = test_set[user,:].toarray()[0,zero_inds].reshape(-1) 
        # Select the binarized yes/no interaction pairs from the original full data
        # that align with the same pairs in training 
        pop = pop_items[zero_inds] # Get the item popularity for our chosen items
        store_auc.append(auc_score(pred, actual)) # Calculate AUC for the given user and store
        popularity_auc.append(auc_score(pop, actual)) # Calculate AUC using most popular and score
    # End users iteration
    
    return float('%.3f'%np.mean(store_auc)), float('%.3f'%np.mean(popularity_auc))  
   # Return the mean AUC rounded to three decimal places for both test and popularity benchmark

We can now use this function to see how our recommender system is doing. To use this function, we will need to transform our output from the ALS function to csr_matrix format and transpose the item vectors. The original pure Python version output the user and item vectors into the correct format already.

In [51]:
calc_mean_auc(product_train, product_users_altered, 
              [sparse.csr_matrix(user_vecs), sparse.csr_matrix(item_vecs.T)], product_test)
# AUC for our recommender system

C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.

C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.

C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.

C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.

C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.

C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.

C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.

C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.

C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.

C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.

C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.

C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.

C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.

C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.

C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.

C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.

C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.

C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.

C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.

C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.

C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.

C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.

C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.

C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.

C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.

C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.

C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.

C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.

C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.

C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.

C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.

C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.

C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.

C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.

C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.

C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.

C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.

C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.

C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.

C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.

C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)
C:\Users\trevor.muchenje\AppData\Local\Continuum\anaconda3\envs\recsys\lib\site-packages\sklearn\metrics\_ranking.py:943: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
  UndefinedMetricWarning)


(nan, nan)

### A Recommendation Example

We now have our recommender system trained and have proven it beats the benchmark of popularity. An AUC of 0.87 means the system is recommending items the user in fact had purchased in the test set far more frequently than items the user never ended up purchasing. To see an example of how it works, let's examine the recommendations given to a particular user and decide subjectively if they make any sense.

First, however, we need to find a way of retrieving the items already purchased by a user in the training set. Initially, we will create an array of our customers and items we made earlier.

In [52]:
customers_arr = np.array(customers) # Array of customer IDs from the ratings matrix
products_arr = np.array(products) # Array of product IDs from the ratings matrix

Now, we can create a function that will return a list of the item descriptions from our earlier created item lookup table.

In [57]:
def get_items_purchased(customer_id, mf_train, customers_list, products_list, item_lookup):
    '''
    This just tells me which items have been already purchased by a specific user in the training set. 
    
    parameters: 
    
    customer_id - Input the customer's id number that you want to see prior purchases of at least once
    
    mf_train - The initial ratings training set used (without weights applied)
    
    customers_list - The array of customers used in the ratings matrix
    
    products_list - The array of products used in the ratings matrix
    
    item_lookup - A simple pandas dataframe of the unique product ID/product descriptions available
    
    returns:
    
    A list of item IDs and item descriptions for a particular customer that were already purchased in the training set
    '''
    cust_ind = np.where(customers_list == customer_id)[0][0] # Returns the index row of our customer id
    purchased_ind = mf_train[cust_ind,:].nonzero()[1] # Get column indices of purchased items
    prod_codes = products_list[purchased_ind] # Get the stock codes for our purchased items
    return item_lookup.loc[item_lookup.product_id.isin(prod_codes)]

We need to look these up by a customer's ID. Looking at the list of customers:

In [54]:
customers_arr[:5]

array([1, 2, 3, 4, 5], dtype=int64)

In [55]:
products_arr[:5]

array([  8,  11, 118, 122, 125], dtype=int64)

we can see that the first customer listed has an ID of 1. Let's examine their purchases from the training set.

In [58]:
get_items_purchased(1, product_train, customers_arr, products_arr, item_lookup)

product_id                                 data_subscriptions
0           125                             Data Daily Bundle 40MB
1           258  WhatsApp (400MB)+ Pinterest 200MB + 400MB Sasa...
2           140                            Daily Data Bouquet 40MB
3           256              Instagram (20MB) + 20MB Sasai Bouquet
5           179  WhatsApp (20MB)+ Pinterest 10MB + 20MB Sasai B...
6           197                           Data Weekly Bundle 350MB
7           180  WhatsApp (45MB)+ Pinterest 25MB + 45MB Sasai B...
8           118               WhatsApp + Twitter Monthly Bundle $3
9           186                            Daily Data Bouquet 80MB
10          226             Facebook + Instagram Daily Bundle $0.3
15          182  Twitter (45MB)+ LinkedIn 25MB + 45MB Sasai Bou...
16          220  Twitter (140MB)+ LinkedIn 70MB + 140MB Sasai B...
17          236                      Twitter Monthly Balance 400MB
18          294                            Private WiFi Bundle 8GB
22          290                          Data Weekly Bouquet 160MB
24          150                          Private WiFi Bundle  15GB
27          188                              2hr data bundle 1.5GB
29          198                           Data Weekly Bundle 700MB
30            8                           Private WiFi Bundle 75GB
34          156                           Private WiFi Bundle 25GB
36          181            Instagram (400MB) + 400MB Sasai Bouquet
43          155                           Data Weekly Bundle 300MB
46          255  Facebook (45MB)+ SnapChat 25MB + 45MB Sasai Bo...
49          266                WhatsApp + Twitter Weekly Bundle $1
62           11                              Data Daily Bundle 2GB
69          262                            Daily Data Bouquet 20MB
73          142                           Daily Data Bouquet 250MB
105         149                            Data Daily Bundle 600MB

We can see that the customer purchased Bike and bike related item. What items does the recommender system say this customer should purchase? We need to create another function that does this. Let's also import the MinMaxScaler from scikit-learn to help with this.

In [61]:
def rec_items(customer_id, mf_train, user_vecs, item_vecs, customer_list, item_list, item_lookup, num_items = 3):
    '''
    This function will return the top recommended items to our users 
    
    parameters:
    
    customer_id - Input the customer's id number that you want to get recommendations for
    
    mf_train - The training matrix you used for matrix factorization fitting
    
    user_vecs - the user vectors from your fitted matrix factorization
    
    item_vecs - the item vectors from your fitted matrix factorization
    
    customer_list - an array of the customer's ID numbers that make up the rows of your ratings matrix 
                    (in order of matrix)
    
    item_list - an array of the products that make up the columns of your ratings matrix
                    (in order of matrix)
    
    item_lookup - A simple pandas dataframe of the unique product ID/product descriptions available
    
    num_items - The number of items you want to recommend in order of best recommendations. Default is 10. 
    
    returns:
    
    - The top n recommendations chosen based on the user/item vectors for items never interacted with/purchased
    '''
    
    cust_ind = np.where(customer_list == customer_id)[0][0] # Returns the index row of our customer id
    pref_vec = mf_train[cust_ind,:].toarray() # Get the ratings from the training set ratings matrix
    pref_vec = pref_vec.reshape(-1) + 1 # Add 1 to everything, so that items not purchased yet become equal to 1
    pref_vec[pref_vec > 1] = 0 # Make everything already purchased zero
    rec_vector = user_vecs[cust_ind,:].dot(item_vecs.T) # Get dot product of user vector and all item vectors
    # Scale this recommendation vector between 0 and 1
    min_max = MinMaxScaler()
    rec_vector_scaled = min_max.fit_transform(rec_vector.reshape(-1,1))[:,0] 
    recommend_vector = pref_vec*rec_vector_scaled 
    # Items already purchased have their recommendation multiplied by zero
    product_idx = np.argsort(recommend_vector)[::-1][:num_items] # Sort the indices of the items into order 
    # of best recommendations
    rec_list = [] # start empty list to store items
    for index in product_idx:
        code = item_list[index]
        rec_list.append([code, item_lookup.data_subscriptions.loc[item_lookup.product_id == code].iloc[0]]) 
        # Append our descriptions to the list
    codes = [item[0] for item in rec_list]
    descriptions = [item[1] for item in rec_list]
    final_frame = pd.DataFrame({'ProductKey': codes, "Product_Name" : descriptions}) # Create a dataframe 
    return final_frame[['ProductKey', 'Product_Name']] # Switch order of columns around

Essentially, this will retrieve the N highest ranking dot products between our user and item vectors for a particular user. Items already purchased are not recommended to the user. For now, let's use a default of 10 items and see what the recommender system decides to pick for our customer.

In [62]:
rec_items(1, product_train, user_vecs, item_vecs, customers_arr, products_arr, item_lookup,num_items = 3)

ProductKey                           Product_Name
0         263                    1hr data bundle 1GB
1         190  Facebook + Instagram Weekly Bundle $1
2         141                 Daily Data Bouquet 2GB

These recommendations seem quite good! Remember that the recommendation system has no real understanding of what a Bike product is. All it knows is the purchase history. It identified that people purchasing Bike product may also want to purchase bike of a differing use. The recommender system also suggests Hydration Pack, which is bike rider need. I personally was blown away by how well the system seems to pick up on these sorts of shopping patterns. Let's try another user that hasn't made a large number of purchases.

In [63]:
get_items_purchased(29, product_train, customers_arr, products_arr, item_lookup)

product_id                                 data_subscriptions
0           125                             Data Daily Bundle 40MB
1           258  WhatsApp (400MB)+ Pinterest 200MB + 400MB Sasa...
2           140                            Daily Data Bouquet 40MB
5           179  WhatsApp (20MB)+ Pinterest 10MB + 20MB Sasai B...
6           197                           Data Weekly Bundle 350MB
7           180  WhatsApp (45MB)+ Pinterest 25MB + 45MB Sasai B...
9           186                            Daily Data Bouquet 80MB
10          226             Facebook + Instagram Daily Bundle $0.3
15          182  Twitter (45MB)+ LinkedIn 25MB + 45MB Sasai Bou...
20          190              Facebook + Instagram Weekly Bundle $1
21          264                        eLearning Bundle 10GB $1120
22          290                          Data Weekly Bouquet 160MB
27          188                              2hr data bundle 1.5GB
30            8                           Private WiFi Bundle 75GB
34          156                           Private WiFi Bundle 25GB
36          181            Instagram (400MB) + 400MB Sasai Bouquet
38          263                                1hr data bundle 1GB
43          155                           Data Weekly Bundle 300MB
46          255  Facebook (45MB)+ SnapChat 25MB + 45MB Sasai Bo...
49          266                WhatsApp + Twitter Weekly Bundle $1
52          141                             Daily Data Bouquet 2GB
69          262                            Daily Data Bouquet 20MB
105         149                            Data Daily Bundle 600MB

In [64]:
rec_items(29, product_train, user_vecs, item_vecs, customers_arr, products_arr, item_lookup,
                       num_items = 3)

ProductKey                           Product_Name
0         150              Private WiFi Bundle  15GB
1         256  Instagram (20MB) + 20MB Sasai Bouquet
2         236          Twitter Monthly Balance 400MB

It certainly picked up on the bike theme along with accessories items. Again, these recommendations seem very impressive given the system doesn't understand the content behind the recommendations. Let's try one more.

In [65]:
get_items_purchased(12, product_train, customers_arr, products_arr, item_lookup)

product_id                                 data_subscriptions
0           125                             Data Daily Bundle 40MB
1           258  WhatsApp (400MB)+ Pinterest 200MB + 400MB Sasa...
2           140                            Daily Data Bouquet 40MB
5           179  WhatsApp (20MB)+ Pinterest 10MB + 20MB Sasai B...
6           197                           Data Weekly Bundle 350MB
9           186                            Daily Data Bouquet 80MB
17          236                      Twitter Monthly Balance 400MB
18          294                            Private WiFi Bundle 8GB
21          264                        eLearning Bundle 10GB $1120
22          290                          Data Weekly Bouquet 160MB
24          150                          Private WiFi Bundle  15GB
29          198                           Data Weekly Bundle 700MB
30            8                           Private WiFi Bundle 75GB
34          156                           Private WiFi Bundle 25GB
36          181            Instagram (400MB) + 400MB Sasai Bouquet
37          122                           Private WiFi Bundle 50GB
38          263                                1hr data bundle 1GB
43          155                           Data Weekly Bundle 300MB
46          255  Facebook (45MB)+ SnapChat 25MB + 45MB Sasai Bo...
52          141                             Daily Data Bouquet 2GB
69          262                            Daily Data Bouquet 20MB
73          142                           Daily Data Bouquet 250MB
105         149                            Data Daily Bundle 600MB

This customer seems like they are buying products suitable for bike and accessories. What other items does the recommender system think they might like?

In [66]:
rec_items(12, product_train, user_vecs, item_vecs, customers_arr, products_arr, item_lookup,
                       num_items = 3)

ProductKey                                       Product_Name
0         266                WhatsApp + Twitter Weekly Bundle $1
1         220  Twitter (140MB)+ LinkedIn 70MB + 140MB Sasai B...
2         182  Twitter (45MB)+ LinkedIn 25MB + 45MB Sasai Bou...